### PR TITLE
K8SPS-55: Group replication

### PR DESCRIPTION
**Pull request by egegunes**
This commit introduces Group Replication (GR) as a new replication type.

We're using mysql-shell to configure and manage GR. mysql-shell is now included in the operator image and the operator executes mysql-shell commands in its shell. Currently we depend on mysql-shell on following operations:

1. Creating InnoDB cluster
2. Adding new nodes to the cluster
3. Rejoining failed nodes to the cluster
4. Removing nodes from cluster
5. Getting cluster status

If cluster type is GR, the operator won't deploy Orchestrator since the topology is managed by the GR plugin.

On top of MySQL nodes, the operator deploys MySQL Router to handle read-write and read-only traffic. The operator won't create a Service to route traffic to primary pod, all routing will be done in MySQL Router itself. Users need to use the router service with specific port numbers.

Currently, **we don't support switching from async to GR and vice versa.**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/config.log) |
| init-deploy | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/monitoring.log) |
| semi-sync | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/semi-sync.log) |
| service-per-pod | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/service-per-pod.log) |
| scaling | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/users.log) |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/4819e8d/demand-backup.log) |
| We run 9 out of 9|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/4819e8d50102ef9956dc71be4d529d1a2638920a
image: `perconalab/percona-server-mysql-operator:PR-91-4819e8d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/config.log) |
| init-deploy | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/monitoring.log) |
| semi-sync | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/semi-sync.log) |
| service-per-pod | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/service-per-pod.log) |
| scaling | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/users.log) |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/3b7c24d/demand-backup.log) |
| We run 9 out of 9|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/3b7c24d8266795307afc1f327f18f71525da9bcb
image: `perconalab/percona-server-mysql-operator:PR-91-3b7c24d`

**Comment by egegunes** [comment]
MySQL Router PR: https://github.com/percona/percona-docker/pull/628

File: build/ps-entrypoint.sh [hunk 1]
```diff
@@ -309,6 +315,7 @@ if [ "$1" = 'mysqld' -a -z "$wantHelp" ]; then
 
 			CREATE USER 'replication'@'%' IDENTIFIED BY '${REPLICATION_PASSWORD}';
 			GRANT SYSTEM_USER, REPLICATION SLAVE ON *.* to 'replication'@'%';
+			GRANT SELECT ON performance_schema.threads to 'replication'@'%';
```
Comment (hors) [inline build/ps-entrypoint.sh:318]: Can we load it only in case if GR is enabled ?

Comment (hors) [reply build/ps-entrypoint.sh:318]: And for backups we need to add GROUP_REPLICATION_ADMIN for xtrabackup user.

File: deploy/cr.yaml [hunk 2]
```diff
@@ -51,7 +51,14 @@ spec:
           bucket: S3-BACKUP-BUCKET-NAME-HERE
           credentialsSecret: cluster1-s3-credentials
           region: us-west-2
+
+  router:
+    image: perconalab/percona-server-mysql-operator:router
```
Comment (hors) [inline deploy/cr.yaml:56]: we need to have affinity and resources sections for router

File: deploy/cr.yaml [hunk 3]
```diff
@@ -121,7 +128,7 @@ spec:
     image: perconalab/percona-server-mysql-operator:main-orchestrator
     imagePullPolicy: Always
 
-    size: 3
+    size: 0
```
Comment (hors) [inline deploy/cr.yaml:131]: can we check that `clusterType: async` and do not start orchestrator even if user change it to 3

File: pkg/mysqlsh/mysqlsh.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "rejoin instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) RemoveInstance(ctx context.Context, clusterName, instance string) error {
+	cmd := fmt.Sprintf("dba.getCluster('%s').removeInstance('%s', {'interactive': false, 'force': true})", clusterName, instance)
```
Comment (hors) [inline pkg/mysqlsh/mysqlsh.go:99]: Why do we need to have `'force': true`?

File: pkg/mysqlsh/mysqlsh.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+func (m *mysqlsh) AddInstance(ctx context.Context, clusterName, instance, podIp string) error {
+	opts := struct {
+		Interactive    bool   `json:"interactive"`
+		RecoveryMethod string `json:"recoveryMethod"`
+		WaitRecovery   int    `json:"waitRecovery"`
+		IpAllowList    string `json:"ipAllowList"`
+	}{
+		Interactive:    false,
+		RecoveryMethod: "clone",
```
Comment (hors) [inline pkg/mysqlsh/mysqlsh.go:69]: Can we use auto?
`auto: let Group Replication choose whether or not a full snapshot has to be taken, based on what the target server supports and the group_replication_clone_threshold sysvar. This is the default value. A prompt will be shown if not possible to automatically determine a convenient way forward. If the interaction is disabled, the operation will be cancelled instead.`

You can create separate task for research,

**Review by hors — CHANGES_REQUESTED** [review]

File: pkg/mysqlsh/mysqlsh.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+	"github.com/percona/percona-server-mysql-operator/pkg/innodbcluster"
+)
+
+type mysqlsh struct {
+	uri  string
+	exec k8sexec.Interface
+}
+
+var sensitiveRegexp = regexp.MustCompile(":.*@")
```
Comment (pooknull) [inline pkg/mysqlsh/mysqlsh.go:22]: I think this regexp can cause problems while printing error. Does mysqlsh really prints password in stdout/stderr?

File: pkg/mysqlsh/mysqlsh.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+	args := []string{"--result-format", "json", "--uri", m.uri, "--cluster", "--", "cluster", "status"}
+
+	c := m.exec.CommandContext(ctx, "mysqlsh", args...)
+	c.SetStdout(&outb)
+	c.SetStderr(&errb)
+
+	status := innodbcluster.Status{}
+
+	if err := c.Run(); err != nil {
```
Comment (pooknull) [inline pkg/mysqlsh/mysqlsh.go:135]: I think we can use a `m.run` method here if we change it to `func (m *mysqlsh) run(ctx context.Context, cmd string, ...args string) error`

**Review by pooknull — CHANGES_REQUESTED** [review]

File: api/v1alpha1/perconaservermysql_types.go [hunk 8]
```diff
@@ -200,6 +201,12 @@ type BackupStorageAzureSpec struct {
 	StorageClass string `json:"storageClass,omitempty"`
 }
 
+type MySQLRouterSpec struct {
+	Expose ServiceExpose `json:"expose,omitempty"`
```
Comment (hors) [inline api/v1alpha1/perconaservermysql_types.go:205]:
```suggestion
	Expose ServiceExposeTogglable `json:"expose,omitempty"`
```

File: build/ps-entrypoint.sh (hunk 1 above)
Comment (egegunes) [reply build/ps-entrypoint.sh:318]: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/037bb02d3f7b8602ac82cf519aa75d2217a61b20

File: deploy/cr.yaml (hunk 2 above)
Comment (egegunes) [reply deploy/cr.yaml:56]: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/7a939c9b4b475549bd2f461f2334c07ebf35a249

File: cmd/manager/main.go [hunk 9]
```diff
@@ -133,6 +136,10 @@ func main() {
 		os.Exit(1)
 	}
 
+	go func() {
+		log.Println(http.ListenAndServe(":6000", nil))
+	}()
+
```
[the remark this answers is not in the document]
Comment (egegunes) [reply cmd/manager/main.go]: no, we don't:  https://github.com/percona/percona-server-mysql-operator/pull/91/commits/3f9149d962ffd9f45d73902be6df02cccc1924db

**Comment by egegunes** [comment]
> we do not have orc but it is ready :)
> 
> ```
> ❯ kubectl get pods
> NAME                                            READY   STATUS    RESTARTS   AGE
> cluster1-mysql-0                                2/2     Running   0          115s
> cluster1-mysql-1                                2/2     Running   0          62s
> cluster1-mysql-2                                0/2     Pending   0          5s
> cluster1-router-848f77cc74-9pj2n                1/1     Running   0          55s
> cluster1-router-848f77cc74-q5s6s                1/1     Running   0          55s
> cluster1-router-848f77cc74-x499p                1/1     Running   0          55s
> percona-server-mysql-operator-f68bb9f46-46kkp   1/1     Running   0          24m
> ❯ kubectl get ps
> NAME       ENDPOINT   MYSQL          ORCHESTRATOR   STATE          AGE
> cluster1              initializing   ready          initializing   2m7s
> ```
> 
> and we do not state foe router and ENDPOINT

https://github.com/percona/percona-server-mysql-operator/pull/91/commits/f642f344c7171c570df97c3697aa025ad4ecc9fc

File: deploy/cr.yaml (hunk 3 above)
Comment (egegunes) [reply deploy/cr.yaml]: we already do it: https://github.com/percona/percona-server-mysql-operator/pull/91/files#diff-91ee13a891cd7d20c59089733ada15406d28a5249194c7922b7e6e9fc1e96ac1R604-R606

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/semi-sync.log) |
| service-per-pod | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/users.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/492bdd7/demand-backup.log) |
| We run 9 out of 9|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/492bdd7858b76c1a8ff2f756718e6d6ddf36bc44
image: `perconalab/percona-server-mysql-operator:PR-91-492bdd7`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/semi-sync.log) |
| service-per-pod | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/sidecars.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/users.log) |
| groupreplication | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/groupreplication.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/a3b6eef/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/a3b6eefa4e5e2b96d71660097a466c86df7f8c84
image: `perconalab/percona-server-mysql-operator:PR-91-a3b6eef`

File: api/v1alpha1/perconaservermysql_types.go [hunk 10]
```diff
@@ -200,6 +209,12 @@ type BackupStorageAzureSpec struct {
 	StorageClass string `json:"storageClass,omitempty"`
 }
 
+type MySQLRouterSpec struct {
+	Expose ServiceExposeTogglable `json:"expose,omitempty"`
```
Comment (hors) [inline api/v1alpha1/perconaservermysql_types.go:213]: we do not need enable option for router

**Review by hors — CHANGES_REQUESTED** [review]
please create the task to improve it in next release 
```
❯ kubectl get ps
NAME       REPLICATION         ENDPOINT        STATE   AGE
cluster1   group-replication   34.159.123.78   ready   23m
```

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/init-deploy.log) |
| monitoring | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/semi-sync.log) |
| service-per-pod | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/users.log) |
| groupreplication | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/groupreplication.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/e62809d/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/e62809d455c4b3862a25fc387479af3bead92fee
image: `perconalab/percona-server-mysql-operator:PR-91-e62809d`

**Review by pooknull — CHANGES_REQUESTED** [review]

File: api/v1alpha1/perconaservermysql_types.go (hunk 10 above)
Comment (egegunes) [reply api/v1alpha1/perconaservermysql_types.go]: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/862d03e22c696064b7de939b4948503e61b5f6bf

File: controllers/perconaservermysql_controller.go [hunk 11]
```diff
@@ -881,33 +1096,64 @@ func (r *PerconaServerMySQLReconciler) reconcileCRStatus(
...
+			(cr.Spec.Router.Expose.Enabled && cr.Spec.Router.Expose.Type != corev1.ServiceTypeLoadBalancer) {
+			return serviceName + "." + cr.GetNamespace(), nil
+		}
+	}
+
+	if cr.Spec.MySQL.IsAsync() {
+		serviceName := mysql.PrimaryServiceName(cr)
+		if cr.Spec.MySQL.PrimaryServiceType != corev1.ServiceTypeLoadBalancer {
+			return serviceName + "." + cr.GetNamespace(), nil
+		}
```
[the remark this answers is not in the document]
Comment (egegunes) [reply controllers/perconaservermysql_controller.go:1155]: I think the current defaults are enough for router, if user doesn't specify expose section the service type will be `ClusterIP`. 

For the validation I thought openapi validation will work but apparently it's not. But operator throws an error when it tries to reconcile service: `reconcile: services: reconcile router svc: patch ps/cluster1-router: Service "cluster1-router" is invalid: spec.type: Unsupported value: "Infinity": supported        values: "ClusterIP", "ExternalName", "LoadBalancer", "NodePort"`. IMO this error is helpful enough, what do you think?

File: pkg/mysqlsh/mysqlsh.go (hunk 6 above)
Comment (egegunes) [reply pkg/mysqlsh/mysqlsh.go:22]: it's not the `mysql-shell` but us that print password to stdout/stderr (without this regexp). We pass instance URI to various commands in `mysqlsh` and we print the command and its arguments if there's an error. This URI has password in most cases.

File: pkg/mysqlsh/mysqlsh.go (hunk 7 above)
Comment (egegunes) [reply pkg/mysqlsh/mysqlsh.go:135]: the reason of duplication here is not `args` but the unmarshaling the command output. I agree it's not optimal, but I intend to keep it as it is right now. What do you think?

File: pkg/mysqlsh/mysqlsh.go (hunk 5 above)
Comment (egegunes) [reply pkg/mysqlsh/mysqlsh.go:69]: https://jira.percona.com/browse/K8SPS-101

**Review by hors — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/users.log) |
| groupreplication | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/groupreplication.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-91/862d03e/demand-backup.log) |
| We run 10 out of 10|

commit: https://github.com/percona/percona-server-mysql-operator/pull/91/commits/862d03e22c696064b7de939b4948503e61b5f6bf
image: `perconalab/percona-server-mysql-operator:PR-91-862d03e`

File: pkg/mysqlsh/mysqlsh.go (hunk 7 above)
Comment (pooknull) [reply pkg/mysqlsh/mysqlsh.go:135]: Missed the unmarshalling part. I think it's ok

File: controllers/perconaservermysql_controller.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pooknull) [reply controllers/perconaservermysql_controller.go:1155]: I mean if we provide empty or invalid `clusterType`, there will be an error about trying to find service with empty name

[the remark this answers is not in the document]
Comment (hors) [reply controllers/perconaservermysql_controller.go:1155]: yes, we will have
` 2022-06-24T09:41:24.089Z	ERROR	controller.perconaservermysql.PerconaServerMySQL	failed to update status	{"reconciler group": "ps.percona.com", "reconciler kind": "PerconaServerMySQL", "name": "cluster1", "namespace": "default", "error": "get app host: get  service: Service \"\" not found", "errorVerbose": "Service \"\" not found\nget  service\ngithub.com/percona/percona-server-mysql-operator/controllers.appHost\n\t/go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysql_controller.go:1161`

And operator will try to update cluster but is stuck

[the remark this answers is not in the document]
Comment (hors) [reply controllers/perconaservermysql_controller.go:1155]: @pooknull I suggest to fix it under separate PR

**Pull request merged**